---
title: "Compare pandora and custom coverage breadth approaches for calling genes as present/absent"
author: "Gavin Douglas"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
---


```{r setup, include=FALSE}
library(ggplot2)
library(reshape2)
library(kableExtra)
library(knitr)

distinct_21_colours <- c("#5b531d",
                          "#874bd2",
                          "#8aad00",
                          "#b03bc1",
                          "#ef8e00",
                          "#9387ff",
                          "#837200",
                          "#b7008b",
                          "#01d1bd",
                          "#ff3a88",
                          "#92d5a3",
                          "#a5102d",
                          "#016cba",
                          "#ff5b4b",
                          "#5e4298",
                          "#cbca7d",
                          "#8c2d6e",
                          "#c9a275",
                          "#fea9ff",
                          "#89374a",
                          "#ff8d8a")
```

# Compare inferred presence / absence profiles

```{r read_in}
pandora_output <- read.table("/data1/gdouglas/projects/honey_bee/combined_Ellegaard.2019.2020/pandora_running/pandora_multisample.matrix",
                             header = TRUE, sep = "\t", row.names = 1)
rownames(pandora_output) <- gsub("\\.fa$", "", rownames(pandora_output))

# Need to revert to old "Gilli_" and "Snod_" format for this...
rownames(pandora_output) <- gsub("Gilliamella_", "Gilli_", rownames(pandora_output))
rownames(pandora_output) <- gsub("Snodgrassella_", "Snod_", rownames(pandora_output))

breadth_out <- readRDS("/data1/gdouglas/projects/honey_bee/combined_Ellegaard.2019.2020/panaroo_breadth_coverage/Gilliamella_and_Snodgrassella_present.rds")

breadth_out_0.1 <- breadth_out
breadth_out_0.5 <- breadth_out
breadth_out_0.9 <- breadth_out

breadth_out_0.1[breadth_out_0.1 < 0.1] <- 0
breadth_out_0.1[breadth_out_0.1 >= 0.1] <- 1
breadth_out_0.1 <- breadth_out_0.1[-which(rowSums(breadth_out_0.1) == 0), ]

breadth_out_0.5[breadth_out_0.5 < 0.5] <- 0
breadth_out_0.5[breadth_out_0.5 >= 0.5] <- 1
breadth_out_0.5 <- breadth_out_0.5[-which(rowSums(breadth_out_0.5) == 0), ]

breadth_out_0.9[breadth_out_0.9 < 0.9] <- 0
breadth_out_0.9[breadth_out_0.9 >= 0.9] <- 1
breadth_out_0.9 <- breadth_out_0.9[-which(rowSums(breadth_out_0.9) == 0), ]

# length(grep("Gilli_", rownames(pandora_output)))
# length(grep("Gilli_", rownames(breadth_out_0.1)))
# length(grep("Gilli_", rownames(breadth_out_0.5)))
# length(grep("Gilli_", rownames(breadth_out_0.9)))

intersecting_genes <- rownames(pandora_output)[which(rownames(pandora_output) %in% rownames(breadth_out_0.5))]
intersecting_genes_Gilli <- intersecting_genes[grep("Gilli_", intersecting_genes)]
intersecting_genes_Snod <- intersecting_genes[grep("Snod_", intersecting_genes)]
```


## Compare observed and expected intersect of positive calls

```{r obs_vs_exp_positive_calls, fig.width=8, fig.height=8}
Gilli_obs <- c()
Gilli_exp <- c()

Snod_obs <- c()
Snod_exp <- c()

for (SRR in colnames(pandora_output)) {
  pandora_Gilli_present_calls <- which(pandora_output[intersecting_genes_Gilli, SRR] == 1)
  breadth_Gilli_present_calls <- which(breadth_out_0.5[intersecting_genes_Gilli, SRR] == 1)
  Gilli_obs <- c(Gilli_obs, length(which(pandora_Gilli_present_calls %in% breadth_Gilli_present_calls)))
  Gilli_exp <- c(Gilli_exp, length(pandora_Gilli_present_calls) * (length(breadth_Gilli_present_calls) / length(intersecting_genes_Gilli)))
  
  pandora_Snod_present_calls <- which(pandora_output[intersecting_genes_Snod, SRR] == 1)
  breadth_Snod_present_calls <- which(breadth_out_0.5[intersecting_genes_Snod, SRR] == 1)
  Snod_obs <- c(Snod_obs, length(which(pandora_Snod_present_calls %in% breadth_Snod_present_calls)))
  Snod_exp <- c(Snod_exp, length(pandora_Snod_present_calls) * (length(breadth_Snod_present_calls) / length(intersecting_genes_Snod)))
  
}

par(mfrow=c(2,1))
plot(Gilli_obs, Gilli_exp, main = "Gilliamella", xlab = "Observed intersecting positive calls", ylab = "Expected intersecting positive calls at random")
abline(b = 1, a = 0, lty = 2, lwd = 2)

plot(Snod_obs, Snod_exp, main = "Snodgrassella", xlab = "Observed intersecting positive calls", ylab = "Expected intersecting positive calls at random")
abline(b = 1, a = 0, lty = 2, lwd = 2)
par(mfrow=c(1,1))
```


## Compare total numbers of positive calls per sample.

```{r compare_num_pos, fig.width=8, fig.height=8}
Gilli_pandora_counts <- c()
Gilli_breadth_counts <- c()

Snod_pandora_counts <- c()
Snod_breadth_counts <- c()

for (SRR in colnames(pandora_output)) {
  Gilli_pandora_counts <- c(Gilli_pandora_counts, length(which(pandora_output[intersecting_genes_Gilli, SRR] == 1)))
  Gilli_breadth_counts <- c(Gilli_breadth_counts, length(which(breadth_out_0.5[intersecting_genes_Gilli, SRR] == 1)))
  
  Snod_pandora_counts <- c(Snod_pandora_counts, length(which(pandora_output[intersecting_genes_Snod, SRR] == 1)))
  Snod_breadth_counts <- c(Snod_breadth_counts, length(which(breadth_out_0.5[intersecting_genes_Snod, SRR] == 1)))
}

par(mfrow=c(2,1))
plot(Gilli_pandora_counts, Gilli_breadth_counts, main = "Gilliamella", xlab = "# pandora positive calls", ylab = "# breadth (50%) positive calls",
     xlim = c(0, 2250), ylim = c(0, 2250))
abline(b = 1, a = 0, lty = 2, lwd = 2)

plot(Snod_pandora_counts, Snod_breadth_counts, main = "Snodgrassella", xlab = "# pandora positive calls", ylab = "# breadth (50%) positive calls",
xlim = c(0, 1500), ylim = c(0, 1500))
abline(b = 1, a = 0, lty = 2, lwd = 2)
par(mfrow=c(1,1))
```


## For one sample (SRR7287243) compare depth and coverage of genes that intersect vs those that don't

```{r one_sample_depth_compare, fig.width = 10, fig.height = 6}
Gilli_mean_depth <- readRDS("/data1/gdouglas/projects/honey_bee/combined_Ellegaard.2019.2020/panaroo_breadth_coverage/Gilliamella_mean_depth_per_site.rds")

SRR <- "SRR7287243"

pandora_pos_genes <- intersecting_genes_Gilli[which(pandora_output[intersecting_genes_Gilli, SRR] == 1)]
breadth0.5_pos_genes <- intersecting_genes_Gilli[which(breadth_out_0.5[intersecting_genes_Gilli, SRR] == 1)]

pandora_only_genes <- pandora_pos_genes[which(! pandora_pos_genes %in% breadth0.5_pos_genes)]
breadth0.5_only_genes <- breadth0.5_pos_genes[which(! breadth0.5_pos_genes %in% pandora_pos_genes)]
both_genes <- pandora_pos_genes[which(pandora_pos_genes %in% breadth0.5_pos_genes)]

par(mfrow = c(2, 2))
boxplot(Gilli_mean_depth[pandora_only_genes, SRR], Gilli_mean_depth[breadth0.5_only_genes, SRR], Gilli_mean_depth[both_genes, SRR],
        main = "Gilli (sample SRR7287243)", names = c("pandora only", "breadth0.5 only", "both"), ylab = "Mean depth per site")

boxplot(breadth_out[pandora_only_genes, SRR], breadth_out[breadth0.5_only_genes, SRR], breadth_out[both_genes, SRR],
        main = "Gilli (sample SRR7287243)", names = c("pandora only", "breadth0.5 only", "both"), ylab = "Breadth of coverage")


Snod_mean_depth <- readRDS("/data1/gdouglas/projects/honey_bee/combined_Ellegaard.2019.2020/panaroo_breadth_coverage/Snodgrassella_mean_depth_per_site.rds")

pandora_pos_genes <- intersecting_genes_Snod[which(pandora_output[intersecting_genes_Snod, SRR] == 1)]
breadth0.5_pos_genes <- intersecting_genes_Snod[which(breadth_out_0.5[intersecting_genes_Snod, SRR] == 1)]

pandora_only_genes <- pandora_pos_genes[which(! pandora_pos_genes %in% breadth0.5_pos_genes)]
breadth0.5_only_genes <- breadth0.5_pos_genes[which(! breadth0.5_pos_genes %in% pandora_pos_genes)]
both_genes <- pandora_pos_genes[which(pandora_pos_genes %in% breadth0.5_pos_genes)]


boxplot(Snod_mean_depth[pandora_only_genes, SRR], Snod_mean_depth[breadth0.5_only_genes, SRR], Snod_mean_depth[both_genes, SRR],
        main = "Snod (sample SRR7287243)", names = c("pandora only", "breadth0.5 only", "both"), ylab = "Mean depth per site")

boxplot(breadth_out[pandora_only_genes, SRR], breadth_out[breadth0.5_only_genes, SRR], breadth_out[both_genes, SRR],
        main = "Snod (sample SRR7287243)", names = c("pandora only", "breadth0.5 only", "both"), ylab = "Breadth of coverage")
par(mfrow = c(1, 1))
```


# Check which method better identifies core genes as present {.tabset}

## Bifidobacterium

```{r Bifidobacterium_core_gene_evaluate, fig.width=6, fig.height=6}
Bifidobacterium_path_to_panaroo <- "/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/panaroo_out_ref_only/Bifidobacterium/gene_presence_absence_roary.csv"

Bifidobacterium_panaroo_out <- read.table(Bifidobacterium_path_to_panaroo,
                                        header = TRUE, sep = ",", stringsAsFactors = FALSE, quote = "", comment.char = "", row.names = 1)
rownames(Bifidobacterium_panaroo_out) <- paste("Bifidobacterium", rownames(Bifidobacterium_panaroo_out), sep = "_")

Bifidobacterium_panaroo_core_genes <- rownames(Bifidobacterium_panaroo_out)[which(Bifidobacterium_panaroo_out$No..isolates >= 14)]
prop_core_gene <- colSums(pandora_output[Bifidobacterium_panaroo_core_genes, ]) / colSums(pandora_output[grep("Bifidobacterium", rownames(Bifidobacterium_panaroo_out)), ])
#colSums(pandora_output[grep("Bifidobacterium", rownames(Bifidobacterium_panaroo_out)), ])

plot(colSums(pandora_output[grep("Bifidobacterium", rownames(Bifidobacterium_panaroo_out)), ]),
     colSums(pandora_output[Bifidobacterium_panaroo_core_genes, ]),
     ylim = c(0, 1200),
     xlim = c(0, 1500),
     xlab = "Total Bifidobacterium genes",
     ylab = "Total core genes identified",
     main = "pandora - Bifidobacterium")
abline(a = length(Bifidobacterium_panaroo_core_genes), b = 0, lwd = 1, lty = 2)
```


## Firm4

```{r Firm4_core_gene_evaluate, fig.width=6, fig.height=6}
Firm4_path_to_panaroo <- "/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/panaroo_out_ref_only/Firm4/gene_presence_absence_roary.csv.gz"

Firm4_panaroo_out <- read.table(Firm4_path_to_panaroo,
                                          header = TRUE, sep = ",", stringsAsFactors = FALSE, quote = "", comment.char = "", row.names = 1)
rownames(Firm4_panaroo_out) <- paste("Firm4", rownames(Firm4_panaroo_out), sep = "_")

Firm4_panaroo_core_genes <- rownames(Firm4_panaroo_out)[which(Firm4_panaroo_out$No..isolates >= 4)]
prop_core_gene <- colSums(pandora_output[Firm4_panaroo_core_genes, ]) / colSums(pandora_output[grep("Firm4", rownames(Firm4_panaroo_out)), ])
#colSums(pandora_output[grep("Firm4", rownames(Firm4_panaroo_out)), ])

plot(colSums(pandora_output[grep("Firm4", rownames(Firm4_panaroo_out)), ]),
     colSums(pandora_output[Firm4_panaroo_core_genes, ], na.rm = TRUE),
     ylim = c(0, 325),
     xlim = c(0, 4000),
     xlab = "Total Firm4 genes",
     ylab = "Total (very soft) core genes identified",
     main = "pandora - Firm4 (VERY SOFT)")
abline(a = length(Firm4_panaroo_core_genes), b = 0, lwd = 1, lty = 2)
```

## Firm5

```{r Firm5_core_gene_evaluate, fig.width=6, fig.height=6}
Firm5_path_to_panaroo <- "/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/panaroo_out_ref_only//Firm5/gene_presence_absence_roary.csv"

Firm5_panaroo_out <- read.table(Firm5_path_to_panaroo,
                                header = TRUE, sep = ",", stringsAsFactors = FALSE, quote = "", comment.char = "", row.names = 1)
rownames(Firm5_panaroo_out) <- paste("Firm5", rownames(Firm5_panaroo_out), sep = "_")

Firm5_panaroo_core_genes <- rownames(Firm5_panaroo_out)[which(Firm5_panaroo_out$No..isolates >= 25)]
prop_core_gene <- colSums(pandora_output[Firm5_panaroo_core_genes, ]) / colSums(pandora_output[grep("Firm5", rownames(Firm5_panaroo_out)), ])
#colSums(pandora_output[grep("Firm5", rownames(Firm5_panaroo_out)), ])

plot(colSums(pandora_output[grep("Firm5", rownames(Firm5_panaroo_out)), ]),
     colSums(pandora_output[Firm5_panaroo_core_genes, ]),
     ylim = c(0, 300),
     xlim = c(0, 4500),
     xlab = "Total Firm5 genes",
     ylab = "Total core genes identified",
     main = "pandora - Firm5")
abline(a = length(Firm5_panaroo_core_genes), b = 0, lwd = 1, lty = 2)
```

## Gilliamella

```{r Gilliamella_core_gene_evaluate, fig.width=10, fig.height=4}
Gilliamella_path_to_panaroo <- "/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/panaroo_out_ref_only/Gilliamella/gene_presence_absence_roary.csv"

Gilliamella_panaroo_out <- read.table(Gilliamella_path_to_panaroo,
                                      header = TRUE, sep = ",", stringsAsFactors = FALSE, quote = "", comment.char = "", row.names = 1)
rownames(Gilliamella_panaroo_out) <- paste("Gilli", rownames(Gilliamella_panaroo_out), sep = "_")

Gilliamella_panaroo_core_genes <- rownames(Gilliamella_panaroo_out)[which(Gilliamella_panaroo_out$No..isolates >= 99)]
prop_core_gene <- colSums(pandora_output[Gilliamella_panaroo_core_genes, ]) / colSums(pandora_output[grep("Gilli", rownames(Gilliamella_panaroo_out)), ])
#colSums(pandora_output[grep("Gilli", rownames(Gilliamella_panaroo_out)), ])

breadth_core_genes_present <- Gilliamella_panaroo_core_genes[which(Gilliamella_panaroo_core_genes %in% rownames(breadth_out_0.5))]
colSums(breadth_out_0.5[breadth_core_genes_present, ])

par(mfrow = c(1, 2))
plot(colSums(pandora_output[grep("Gilli", rownames(Gilliamella_panaroo_out)), ]),
     colSums(pandora_output[Gilliamella_panaroo_core_genes, ]),
     ylim = c(0, 700),
     xlim = c(0, 9000),
     xlab = "Total Gilliamella genes",
     ylab = "Total core genes identified",
     main = "pandora")
abline(a = length(Gilliamella_panaroo_core_genes), b = 0, lwd = 1, lty = 2)


plot(colSums(breadth_out_0.5[grep("Gilli", rownames(breadth_out_0.5)), ]),
     colSums(breadth_out_0.5[breadth_core_genes_present, ]),
     ylim = c(0, 700),
     xlim = c(0, 9000),
     xlab = "Total Gilliamella genes",
     ylab = "Total core genes identified",
     main = "breadth (0.5)")
abline(a = length(Gilliamella_panaroo_core_genes), b = 0, lwd = 1, lty = 2)

par(mfrow = c(1, 1))
```

## Snodgrassella

```{r Snodgrassella_core_gene_evaluate, fig.width=10, fig.height=4}
Snodgrassella_path_to_panaroo <- "/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/panaroo_out_ref_only/Snodgrassella/gene_presence_absence_roary.csv"

Snodgrassella_panaroo_out <- read.table(Snodgrassella_path_to_panaroo,
                                      header = TRUE, sep = ",", stringsAsFactors = FALSE, quote = "", comment.char = "", row.names = 1)
rownames(Snodgrassella_panaroo_out) <- paste("Snod", rownames(Snodgrassella_panaroo_out), sep = "_")

Snodgrassella_panaroo_core_genes <- rownames(Snodgrassella_panaroo_out)[which(Snodgrassella_panaroo_out$No..isolates >= 51)]
prop_core_gene <- colSums(pandora_output[Snodgrassella_panaroo_core_genes, ]) / colSums(pandora_output[grep("Snod", rownames(Snodgrassella_panaroo_out)), ])
#colSums(pandora_output[grep("Snod", rownames(Snodgrassella_panaroo_out)), ])

breadth_core_genes_present <- Snodgrassella_panaroo_core_genes[which(Snodgrassella_panaroo_core_genes %in% rownames(breadth_out_0.5))]
#colSums(breadth_out_0.5[breadth_core_genes_present, ])

par(mfrow = c(1, 2))
plot(colSums(pandora_output[grep("Snod", rownames(Snodgrassella_panaroo_out)), ]),
     colSums(pandora_output[Snodgrassella_panaroo_core_genes, ]),
     ylim = c(0, 1300),
     xlim = c(0, 4000),
     xlab = "Total Snodgrassella genes",
     ylab = "Total core genes identified",
     main = "pandora")
abline(a = length(Snodgrassella_panaroo_core_genes), b = 0, lwd = 1, lty = 2)


plot(colSums(breadth_out_0.5[grep("Snod", rownames(breadth_out_0.5)), ]),
     colSums(breadth_out_0.5[breadth_core_genes_present, ]),
     ylim = c(0, 1300),
     xlim = c(0, 4000),
     xlab = "Total Snodgrassella genes",
     ylab = "Total core genes identified",
     main = "breadth (0.5)")
abline(a = length(Snodgrassella_panaroo_core_genes), b = 0, lwd = 1, lty = 2)

par(mfrow = c(1, 1))
```


# Compare how concordance is associated with PRG metric {.tabset}

It would be interesting if highly discordant present/absent calls were for genes with complicated PRGs. This would indicate that using a representative sequence alone was the problem in this case.

```{r jaccard_func}

jaccard_dist_two_vecs <- function(x, y) {
  both_called <- length(which(x == 1 & y == 1))
  only_x_called <- length(which(x == 1 & y == 0))
  only_y_called <- length(which(x == 0 & y == 1))
  
  return(1 - both_called / (both_called + only_x_called + only_y_called))
}

```

## Gilliamella

```{r Gilliamella_prg_metric_compare}
intersecting_samples_Gilli <- colnames(pandora_output)[which(colnames(pandora_output) %in% colnames(breadth_out_0.5))]

pandora_output_sorted <- pandora_output[intersecting_genes_Gilli, intersecting_samples_Gilli]
breadth_out_0.5_sorted <- breadth_out_0.5[intersecting_genes_Gilli, intersecting_samples_Gilli]

jaccard_dist <- c()

for (Gilli_gene in intersecting_genes_Gilli) {
  jaccard_dist <- c(jaccard_dist, jaccard_dist_two_vecs(x = as.numeric(pandora_output_sorted[Gilli_gene, ]),
                                                        y = as.numeric(breadth_out_0.5_sorted[Gilli_gene, ])))
}

names(jaccard_dist) <- intersecting_genes_Gilli


Gilli_prg_summary <- read.table("/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/ortholog_pandora_prgs/Gilliamella_ortholog_fastas_muscle_aligned/summary.tsv",
                                header = FALSE, sep = "\t", row.names = 1)
rownames(Gilli_prg_summary) <- gsub("\\.fa$", "", rownames(Gilli_prg_summary))
rownames(Gilli_prg_summary) <- gsub("Gilliamella_", "Gilli_", rownames(Gilli_prg_summary))

Gilli_prg_space_counts <- read.table("/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/ortholog_pandora_prgs/Gilliamella_ortholog_fastas_muscle_aligned/Gilliamella_num_space_per_prg.txt",
                                     header = FALSE, sep = " ", row.names = 1)
rownames(Gilli_prg_space_counts) <- gsub("\\.fa.max_nest5.min_match7.prg$", "", rownames(Gilli_prg_space_counts))
rownames(Gilli_prg_space_counts) <- gsub("Gilliamella_", "Gilli_", rownames(Gilli_prg_space_counts))

Gilli_prg_space_counts$V2[which(is.na(Gilli_prg_space_counts$V2))] <- 0

Gilli_prg_and_jaccard <- data.frame(prg_metric = Gilli_prg_summary[intersecting_genes_Gilli, "V4"],
                                    jaccard_dist = jaccard_dist,
                                    space_counts = Gilli_prg_space_counts[intersecting_genes_Gilli, "V2"],
                                    pandora_prev = rowSums(pandora_output_sorted),
                                    breadth_prev = rowSums(breadth_out_0.5_sorted))

Gilli_prg_and_jaccard_common <- Gilli_prg_and_jaccard[which(Gilli_prg_and_jaccard$pandora_prev > 5 & Gilli_prg_and_jaccard$breadth_prev > 5), ]

plot(Gilli_prg_and_jaccard_common$prg_metric, Gilli_prg_and_jaccard_common$jaccard_dist)
```

## Snodgrassella

```{r Snodgrassella_prg_metric_compare}
intersecting_samples_Snod <- colnames(pandora_output)[which(colnames(pandora_output) %in% colnames(breadth_out_0.5))]

pandora_output_sorted <- pandora_output[intersecting_genes_Snod, intersecting_samples_Snod]
breadth_out_0.5_sorted <- breadth_out_0.5[intersecting_genes_Snod, intersecting_samples_Snod]

jaccard_dist <- c()

for (Snod_gene in intersecting_genes_Snod) {
  jaccard_dist <- c(jaccard_dist, jaccard_dist_two_vecs(x = as.numeric(pandora_output_sorted[Snod_gene, ]),
                                                        y = as.numeric(breadth_out_0.5_sorted[Snod_gene, ])))
}

names(jaccard_dist) <- intersecting_genes_Snod


Snod_prg_summary <- read.table("/data1/gdouglas/projects/honey_bee/ref_genome_pangenomes/ortholog_pandora_prgs/Snodgrassella_ortholog_fastas_muscle_aligned/summary.tsv",
                                header = FALSE, sep = "\t", row.names = 1)
rownames(Snod_prg_summary) <- gsub("\\.fa$", "", rownames(Snod_prg_summary))
rownames(Snod_prg_summary) <- gsub("Snodgrassella_", "Snod_", rownames(Snod_prg_summary))

Snod_prg_and_jaccard <- data.frame(prg_metric = Snod_prg_summary[intersecting_genes_Snod, "V4"],
                                    jaccard_dist = jaccard_dist,
                                    pandora_prev = rowSums(pandora_output_sorted),
                                    breadth_prev = rowSums(breadth_out_0.5_sorted))

Snod_prg_and_jaccard_common <- Snod_prg_and_jaccard[which(Snod_prg_and_jaccard$pandora_prev > 5 & Snod_prg_and_jaccard$breadth_prev > 5), ]

plot(Snod_prg_and_jaccard_common$prg_metric, Snod_prg_and_jaccard_common$jaccard_dist)
```


# Concordance correlated with overall sample depth

```{r read_sample_depth}
fastq_depth <- c()

for (SRR in colnames(pandora_output)) {
  
  wc_path <- paste("/data1/gdouglas/projects/honey_bee/filtered_fastqs/Ellegaard.2019.2020_cat_wc_files/", SRR, ".fastq_wc.txt", sep = "")
  
  fastq_depth <- c(fastq_depth, read.table(wc_path)$V1)
  
}

names(fastq_depth) <- colnames(pandora_output)

```

## Gilliamella

```{r Gilliamella_depth_correlation}

Gilli_sample_jaccard_dist <- c()

for (SRR in colnames(pandora_output)) {
  Gilli_genes <- grep("Gilli", rownames(pandora_output), value = TRUE)
  Gilli_genes <- Gilli_genes[which(Gilli_genes %in% rownames(breadth_out_0.5))]

  Gilli_sample_jaccard_dist <- c(Gilli_sample_jaccard_dist, jaccard_dist_two_vecs(x = pandora_output[Gilli_genes, SRR], y = breadth_out_0.5[Gilli_genes, SRR]))
}

names(Gilli_sample_jaccard_dist) <- colnames(pandora_output)


plot(Gilli_sample_jaccard_dist, fastq_depth)
cor.test(Gilli_sample_jaccard_dist, fastq_depth, method = "spearman")
```


## Snodgrassella

```{r Snodgrassella_depth_correlation}

Snod_sample_jaccard_dist <- c()

for (SRR in colnames(pandora_output)) {
  Snod_genes <- grep("Snod", rownames(pandora_output), value = TRUE)
  Snod_genes <- Snod_genes[which(Snod_genes %in% rownames(breadth_out_0.5))]

  Snod_sample_jaccard_dist <- c(Snod_sample_jaccard_dist, jaccard_dist_two_vecs(x = pandora_output[Snod_genes, SRR], y = breadth_out_0.5[Snod_genes, SRR]))
}

names(Snod_sample_jaccard_dist) <- colnames(pandora_output)


plot(Snod_sample_jaccard_dist, fastq_depth)
cor.test(Snod_sample_jaccard_dist, fastq_depth, method = "spearman")
```


# Correlation in prevalence of each gene

## Gilliamella

```{r Gilliamella_prev_correlation}
plot(Gilli_prg_and_jaccard_common$pandora_prev, Gilli_prg_and_jaccard_common$breadth_prev)
cor.test(Gilli_prg_and_jaccard_common$pandora_prev, Gilli_prg_and_jaccard_common$breadth_prev, method = "spearman")
```

## Snodgrassella

```{r Snodgrassella_prev_correlation}
plot(Snod_prg_and_jaccard_common$pandora_prev, Snod_prg_and_jaccard_common$breadth_prev)
cor.test(Snod_prg_and_jaccard_common$pandora_prev, Snod_prg_and_jaccard_common$breadth_prev, method = "spearman")
```

